File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/service/server/AddressService.java
package com.antonio.samir.meteoritelandingsspots.service.server;

import android.location.Address;
import android.util.Log;

import com.antonio.samir.meteoritelandingsspots.Application;
import com.antonio.samir.meteoritelandingsspots.ApplicationDebug;
import com.antonio.samir.meteoritelandingsspots.model.Meteorite;
import com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteRepository;
import com.antonio.samir.meteoritelandingsspots.service.repository.database.MeteoriteDao;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;


public class AddressService {

    public static final String TAG = AddressService.class.getSimpleName();
    final static ThreadPoolExecutor executor = new ThreadPoolExecutor(0, 3,
            1L, TimeUnit.MILLISECONDS,
            new LinkedBlockingQueue<Runnable>());

    private final MeteoriteDao mMeteoriteDao;

    public AddressService() {
        mMeteoriteDao = new MeteoriteRepository(ApplicationDebug.getContext()).getAppDatabase().meteoriteDao();
    }

    public void recoverAddress(final Meteorite meteorite, final String recLat, final String recLong) {

        executor.execute(new Runnable() {
            @Override
            public void run() {
                final String address = getAddress(recLat, recLong);
                meteorite.setAddress(address);
                mMeteoriteDao.update(meteorite);
                Log.i(TAG, String.format("Address for id %s recovered", meteorite.getId()));
            }
        });

    }

    private String getAddress(String recLat, String recLong) {
        String addressString = "";
        if (StringUtils.isNoneEmpty(recLat) && StringUtils.isNoneEmpty(recLong)) {
            final Address address = GeoLocationUtil.getAddress(Double.parseDouble(recLat), Double.parseDouble(recLong), Application.getContext());
            if (address != null) {
                final List<String> finalAddress = new ArrayList<>();
                final String city = address.getLocality();
                if (StringUtils.isNoneEmpty(city)) {
                    finalAddress.add(city);
                }

                final String state = address.getAdminArea();
                if (StringUtils.isNoneEmpty(state)) {
                    finalAddress.add(state);
                }

                final String countryName = address.getCountryName();
                if (StringUtils.isNoneEmpty(countryName)) {
                    finalAddress.add(countryName);
                }

                addressString = StringUtils.join(finalAddress, ", ");

            }
        }

        return addressString;
    }
}

File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/service/business/AddressService.kt
package com.antonio.samir.meteoritelandingsspots.service.business

import android.util.Log
import androidx.annotation.StringDef
import androidx.lifecycle.MutableLiveData
import com.antonio.samir.meteoritelandingsspots.service.business.AddressService.Status.Companion.DONE
import com.antonio.samir.meteoritelandingsspots.service.business.AddressService.Status.Companion.LOADING
import com.antonio.samir.meteoritelandingsspots.service.business.model.Meteorite
import com.antonio.samir.meteoritelandingsspots.service.repository.local.MeteoriteRepositoryInterface
import com.antonio.samir.meteoritelandingsspots.util.DefaultDispatcherProvider
import com.antonio.samir.meteoritelandingsspots.util.DispatcherProvider
import com.antonio.samir.meteoritelandingsspots.util.GeoLocationUtilInterface
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.apache.commons.lang3.StringUtils
import java.util.*


class AddressService(
        private val meteoriteRepository: MeteoriteRepositoryInterface,
        private val geoLocationUtil: GeoLocationUtilInterface,
        private val dispatchers: DispatcherProvider = DefaultDispatcherProvider()
) : AddressServiceInterface {

    val TAG = AddressService::class.java.simpleName

    override val status = MutableLiveData<String>()

    @Retention(AnnotationRetention.SOURCE)
    @StringDef(DONE, LOADING)
    annotation class Status {
        companion object {
            const val DONE = "DONE"
            const val LOADING = "LOADING"
        }
    }

    override fun recoveryAddress() {

        GlobalScope.launch(dispatchers.unconfined()) {
            if (status.value == null || status.value === DONE) {
                var meteorites = meteoriteRepository.meteoritesWithOutAddress()

                Log.i(TAG, "recoveryAddress $LOADING")

                status.postValue(LOADING)

                while (meteorites.isNotEmpty()) {
                    recoverAddress(meteorites)
                    meteorites = meteoriteRepository.meteoritesWithOutAddress()
                }

                status.postValue(DONE)

            }

        }

    }

    override suspend fun recoverAddress(list: List<Meteorite>) {
        try {
            list.onEach { meteorite ->
                meteorite.address = getAddressFromMeteorite(meteorite)
            }
            val meteoritesWithoutAddressCount: Int = meteoriteRepository.getMeteoritesWithoutAddressCount()
            Log.i(TAG, "recoveryAddress ${list.size} $meteoritesWithoutAddressCount")

        } catch (e: Exception) {
            Log.e(TAG, "Fail to retrieve address", e)
        }
        meteoriteRepository.updateAll(list)
    }

    override suspend fun recoverAddress(meteorite: Meteorite) = withContext(dispatchers.default()) {
        meteorite.address = getAddressFromMeteorite(meteorite)
        meteoriteRepository.update(meteorite)
    }

    private fun getAddressFromMeteorite(meteorite: Meteorite): String {
        val recLat = meteorite.reclat
        val recLong = meteorite.reclong

        var metAddress = " "
        if (recLat != null && recLong != null) {
            val address = getAddress(recLat.toDouble(), recLong.toDouble())
            metAddress = address ?: " "
        }
        return metAddress
    }


    private fun getAddress(recLat: Double, recLong: Double): String? {
        var addressString: String? = null
        val address = geoLocationUtil.getAddress(recLat, recLong)
        if (address != null) {
            val finalAddress = ArrayList<String>()
            val city = address.locality
            if (StringUtils.isNoneEmpty(city)) {
                finalAddress.add(city)
            }

            val state = address.adminArea
            if (StringUtils.isNoneEmpty(state)) {
                finalAddress.add(state)
            }

            val countryName = address.countryName
            if (StringUtils.isNoneEmpty(countryName)) {
                finalAddress.add(countryName)
            }

            if (finalAddress.isNotEmpty()) {
                addressString = finalAddress.joinToString(", ")
            }

        }

        return addressString
    }
}
